derivation.nix

```diff
@@ -39,17 +47,66 @@ in
     , ...
     }@attrs:
     let
+      # Flag that can be computed from arguments, indicating a workspace was
+      # supplied. Only used in these let bindings.
+      isWorkspace = workspace != null && components != [];
+      # Utility functions
+      forEachConcat = f: xs: concatStringsSep "\n" (map f xs);
+      forEachComponent = f: forEachConcat f components;
+      # Computed values used below that don't loop
       nativeBuildInputs = [
         nodejs
         pnpm
         pkg-config
       ] ++ extraBuildInputs;
+      copyLink =
+        if copyNodeModules
+          then "cp -r"
+          else "ln -s";
+      packageFilesWithoutLockfile =
+        [
+          { name = "package.json"; value = packageJSON; }
+        ] ++ componentPackageJSONs ++ computedNodeModuleSources;
+      computedNodeModuleSources =
+        (if pnpmWorkspaceYaml == null
+          then []
+          else [
+            {name = "pnpm-workspace.yaml"; value = pnpmWorkspaceYaml;}
+          ]
+        ) ++ extraNodeModuleSources;
+      # Computed values that loop over something
+      nodeModulesDirs =
+        if isWorkspace then
+          ["node_modules"] ++ (map (c: "${c}/node_modules") components)
+        else ["node_modules"];
+      buildScripts =
+        if isWorkspace then ''
+          ${forEachComponent (component: ''
+            (cd "${component}" && pnpm run ${script})
```

Comment:
Can `pnpm --recursive run ${script}` work without needing the component list?

Reply:
We need the components list to knowh which `node_modules` directories we need and for install to know what to copy.

I might be able to parse the pnpm-workspace.yaml file and remove the ability to specify the components.

As long as we accept the components list, it makes more sense to just build each component. There is no guarantee the caller gave us all the components. I might be able to rig something up with filters, but that feels more fragile than just accepting directories as arguments and using them as directories.

Reply:
It turned out that using a filter list was less fragile. A component may be needed without having a particular script.

So I went with a filtered `pnpm run --recursive`.